SaraAlert/SaraAlert
# SARAALERT-788: Add data validation to API

# Description
Jira Ticket: [SARAALERT-788](https://tracker.codev.mitre.org/browse/SARAALERT-788)

This PR adds data validation to the API via increased validation on the `Patient` model. The validation is based on the validation that currently happens in the `ImportController`. FHIR (syntax) errors are reported back with a 400 error and validation errors are reported back with a 422 error. Both types of errors are reported using an [OperationOutcome](https://www.hl7.org/fhir/operationoutcome.html). 

# Demo

Sample 400 OperationOutcome:
```
{
    "issue": [
        {
            "severity": "error",
            "code": "processing",
            "diagnostics": "ContactPoint.value: invalid cardinality. Found 2 expected 0..1"
        }
    ],
    "resourceType": "OperationOutcome"
}
```
Sample 422 OperationOutcome:
```
{
    "issue": [
        {
            "severity": "error",
            "code": "processing",
            "diagnostics": "'18-18-2020' is not a valid date for 'Last Date of Exposure', please use the 'YYYY-MM-DD' format"
        },
        {
            "severity": "error",
            "code": "processing",
            "diagnostics": "'not@an@email.com' is not a valid Email Address for 'Email'"
        }
    ],
    "resourceType": "OperationOutcome"
}
```

# Important Changes

`patient.rb`
* Added additional validation of fields used in the API to the model.

`patient_helper.rb`
* Changed `from_*` functions to only extract values from FHIR, not validate them.

`validation_helper.rb`
* Created this file to contain the validation specific constants that previously existed in `ImportExport`.

`api_controller.rb`
* Updated responses to include error messages from FHIR validation and `Patient` model validation.

`app/models/concerns/*`
* Added custom validators for date, email, and phone numbers.

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):

Tests were added to `api_controller_test.rb` and `patient_test.rb` to test the error reporting and validation, respectively. To test manually you can set up an API client and make POST and PUT requests against the `fhir/r4/Patient` endpoint. There is a sample Patient to start with [here](https://github.com/SaraAlert/SaraAlert/blob/788-api-validation/API.md#request-body).

To test FHIR validation:
* Try changing a field that should be a single value to an array
* Try invalid data, for example "foo" in a field expecting a dateTime

There are four types of validation to test with `Patient` model validation
* Inclusion (ex: `preferred-contact-time`)
* Dates (ex: `birthDate`)
* Phone numbers (ex: `telecom.value` if `telecom.system` is `phone`)
* Emails (ex: `telecom.value` if `telecom.system` is `email`)

# Notes

1. This PR does cause there to be duplicate code between the `ImportController` and the validations that have been added to the `Patient` model. We have planned to refactor the `ImportController` to remove this duplicate code, but I think it is good to get eyes on this PR at this point. If we decide this PR can't be merged without first refactoring out that duplicate code, that's understandable, but I'd still like feedback on this PR before tacking that on as well.
2. It is important to consider how adding this validation could impact existing data that is invalid. I'm not entirely sure how we assess that impact, and would like some assistance on that if possible. On a similar note, one thing that could go wrong with this PR is if the validation on the model was somehow more strict than the restrictions placed on users enrolling patients. For example if the front-end definition of a phone number was less strict. I was unable to find any such cases, but I wanted to mention it so those who know the app better than me could keep it in mind.
3. Wasn't sure what branch to target this at, went with `master` but let me know if I should change that.
